examples/shamir-secret-share.masm

```diff
@@ -0,0 +1,150 @@
+#! - Migrate to a more secure random number generation algorithm for polynomial coefficient generation.
+#! - Allow arbitrary secret size of size n.  This would require generation of n polynomials.
+#! - Instead of computing x^i for each iteration, store x^i in a variable and multiply by x to get next exponent.
+
+
+# create secret commitment (hash of secret + rand)
+proc.gen_secret_com
+    # input: [secret, rand]
+    padw          #[0, 0, 0, 0, secret, rand]
+    padw          #[0, 0, 0, 0, 0, 0, 0, 0, secret, rand]
+    drop          #[0, 0, 0, 0, 0, 0, 0, secret, rand]
+    drop          #[0, 0, 0, 0, 0, 0, secret, rand]
+    rphash        #[com_1, com_2, com_3, com_4]
+end
+
+# create k-1 random polynomial coefficients
+proc.gen_poly_coeffs
+    # input [k, rand]
+
+    # generate initial 4 random coefficients from rand
+    swap    #[rand, k]
+    padw    #[0, 0, 0, 0, rand, k]
+    padw    #[0, 0, 0, 0, 0, 0, 0, 0, rand, k]
+    drop    #[0, 0, 0, 0, 0, 0, 0, rand, k]
+    rphash  #[coef_1, coef_2, coef_3, coef_4, k]
+
+    # introduce counter and evaluate if we have enough cofficients
+    push.5  #[counter, coef_1, coef_2, coef_3, coef_4, k]
+    movup.5 #[k, counter, coef_1, coef_2, coef_3, coef_4]
+    dup.1
+    dup.1
+    lt
+
+    # continue creating more coefficients until we have enough (created in 4's)
+    while.true     #[k, counter, diget_1, diget_2, digest_3, digest_4]
+        # generate more cofficients by padding and hashing previous cofficients
+        movdn.5
+        movdn.5
+        dupw       #[digest_1, digest_2, digest_3, digest_4, k, counter, ...]
+        padw       #[0, 0, 0, 0, diget_1, diget_2, digest_3, digest_4, diget_1, diget_2, digest_3, digest_4, k, counter]
+        rphash     #[digest_5, digest_6, digest_7, digest_8, digest_1, digest_2, digest_3, digest_4, k, counter, ...]
+
+        # evaluate if we have enugh coeficients
+        movup.9
+        add.4
+        movup.9
+        dup.1
+        dup.1
+        lt
+    end         #[k, counter, coef_1, ...]  
+    
+    # evaluate if we have too many cofficients
+    dup          #[k, k, counter, coef_1, ...]
+    movdn.2      #[k, counter, k, coef_1, ...]
+    sub          #[diff, k, coef_1, ...]
+    dup          #[diff, diff, k, coef_1, ...]
+    push.0       #[0, diff, diff, k, coef_1, ...]
+    gt           #[1, diff, k, coef_1] 
+
+    # drop unrequired cofficients such that we have k in total
+    while.true #[diff, k, coef_1]
+        movup.2
+        drop
+        sub.1
+        dup
+        push.0
+        gt
+    end
+
+    drop
+end
+
+# evaluate polynomial
+proc.eval_poly
+    #[x, k, coef_1,...]
+
+    # introduce sum variable to calculate y
+    swap    #[k, x, coef_1, ..., secret]
+    push.0  #[sum, k, x, coef_1, ..., secret]
+    push.1  #[1, sum, k, x, coef_1, ..., secret]
+
+    # evaluate polymoial by iterating through terms and summing
+    while.true 
+        dup.2 #[x, sum, k, x, coef_1,..., secret]
```

Comment:
on the first iteration there is a `1` on the top of the stack, so this is really:

```suggestion
        dup.2 #[k, 1, sum, k, x, coef_1,..., secret]
```

I didn't review the rest of the loop, it seems the `1` is not accounted for the ops below, so the stack should be reviewed.

Reply:
I think the 1 is consumed by the while loop condition evaluation.